src/disk_cache.rs

```diff
@@ -0,0 +1,952 @@
+        // to the object store.
+        // TODO: add a warning log here.
+        entry.save_result(get_result).await.ok();
+        Ok(result_meta)
+    }
+
+    // given the range and object size, split the range into parts, and return the part id and the range
+    // inside the part.
+    fn split_range_into_parts(&self, range: Range<usize>) -> Vec<(PartID, Range<usize>)> {
+        let range_aligned = self.align_range(&range, self.part_size);
+        let start_part = range_aligned.start / self.part_size;
+        let end_part = range_aligned.end / self.part_size;
+        let mut parts: Vec<_> = (start_part..end_part)
+            .map(|part_id| {
+                (
+                    part_id,
+                    Range {
+                        start: 0,
+                        end: self.part_size,
+                    },
+                )
+            })
+            .collect();
+        if parts.is_empty() {
+            return vec![];
+        }
+        if let Some(first_part) = parts.first_mut() {
+            first_part.1.start = range.start % self.part_size;
+        }
+        if let Some(last_part) = parts.last_mut() {
+            if range.end % self.part_size != 0 {
+                last_part.1.end = range.end % self.part_size;
+            }
+        }
+        parts
+    }
+
+    /// get from disk if the parts are cached, otherwise start a new GET request.
+    /// the io errors on reading the disk caches will be ignored, just fallback to
+    /// the object store.
+    /// TODO: add metrics to track the cache hit rate here.
+    fn read_part(
+        &self,
+        location: &Path,
+        part_id: PartID,
+        range_in_part: Range<usize>,
+    ) -> BoxFuture<'static, object_store::Result<Bytes>> {
+        let part_size = self.part_size;
+        let object_store = self.object_store.clone();
+        let root_folder = self.root_folder.clone();
+        let location = location.clone();
+        Box::pin(async move {
+            let entry = DiskCacheEntry {
+                root_folder,
+                location: location.clone(),
+                part_size,
+            };
+            if let Ok(Some(bytes)) = entry.read_part(part_id).await {
+                return Ok(bytes.slice(range_in_part));
+            }
+
+            // if the part is not cached, fallback to the object store to get the missing part.
+            // the object stores is expected to return the result whenever the `start` of the range
+            // is not out of the object size.
+            let range = Range {
+                start: part_id * part_size,
+                end: (part_id + 1) * part_size,
+            };
+            let get_result = object_store
+                .get_opts(
+                    &location,
+                    GetOptions {
+                        range: Some(GetRange::Bounded(range)),
+                        ..Default::default()
```

Comment:
Same note here about GetOptions defaults.

Reply:
the interface is not regarded as a transparent proxy to ObjectStore but a customized `CacheableGetOptions`, so i suppose that we do not need to pass the `get_opts` as user passed before, we can pass the `GetOptions` as we need here.

Reply:
Can you say more here? I'm not understanding.